### PR TITLE
Topic/779 vc repeat

### LINKED ISSUE #779: ValueConverter on collection not executed after adding a new item to the collection
# 🐛 Bug Report

A value converter on a collection is not executed when an item is added to the collection, see:

Is executed after adding an item to the collection in au1:
e2e test run: https://youtu.be/75rCPILsyy8 
code: https://github.com/eriklieben/mua-component-library-au1

AU2 
e2e test run: https://youtu.be/-ZwRDa8HlOQ 
code: https://github.com/eriklieben/mua-component-library

See the last e2e test "should find items that are added to the list after searching", after adding the item it is in the collection (bottom list), but the ```repeat.for``` loop with a value, binding doesn't get executed to refresh/update the result. When (by hand) I perform the same and clear the search term and apply it again, the result is correct (because it executed the value converter in between my actions).

It currently directly updates the items in the collection, because I assumed it had something to do with me cloning the collection, wanted to exclude that as a possibility. Left that in, because that might be easier in finding a solution. 
In production, one would probably clone the collection, before applying modifications, but I think/assume if it executed the value converter after a modification to the collection that scenario should also work again.

## 🤔 Expected Behavior

After a collection is modified (item added) the value converter should be executed.

## 😯 Current Behavior

The value converter is not executed after an item was added.

## 💁 Possible Solution

not sure.

## 💻 Code Sample

AU2 code https://github.com/eriklieben/mua-component-library
same logic in AU1 code: https://github.com/eriklieben/mua-component-library-au1

you can run the cypress e2e tests in both code samples to see the behavior.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Aurelia          |dev
| Language         | ts
| Browser          | chrome 78 on windows 10
| Bundler          | webpack
| Operating System | win 10
| NPM/Node/Yarn    | npm 6.10

Let me know if I need to explain it better or more clearly. Thanks!

### PR DESCRIPTION
**Pull request by Sayan751**
# Pull Request

### 🎫 Issues

Closes #779

### CODE REVIEW

**Review by bigopon — COMMENTED** [review]
Nice work 👍 🚀  I have a few questions and a comment on the testing style for the first round.

File: packages/__tests__/3-runtime-html/repeat.vc.bb.spec.ts [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+      } else {
+        [this.name, this.only] = arg;
+      }
+      this.changes = changes;
+    }
+  }
+
+  function* getTestData() {
+    const identityArrTmplt = `<span repeat.for="item of arr | identity">\${item}</span>`;
+    yield new TestData(
```
[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/__tests__/3-runtime-html/repeat.vc.bb.spec.ts:142]: Having a class make it easier to add methods to it. For complex tests, I often tend to add methods to such test data classes.

The test data is very much context-specific; i.e. the shape of the class depends on the nature of the test. Thus the test data for testing the `promise` TC should be sufficiently different from the test data class to test the validation controller. The attempt to unify that might be futile with confusing end result. We may find some commonality between these classes and have a base class in place, but that curbs the flexibility in an undesired way.

On the other hand using object literal means more typing; especially if you are using VS Code:
image

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 2]
```diff
@@ -156,13 +182,29 @@ export class Repeat<C extends Collection = unknown[]> implements ICustomAttribut
 
   // todo: subscribe to collection from inner expression
   private _checkCollectionObserver(flags: LF): void {
+    const scope = this.$controller.scope;
+    let observingInnerItems = this.observingInnerItems;
+    if ((flags & LF.fromBind) > 0) {
+      let expression = this.forOf.iterable;
+      while (expression != null && wrappedExprs.includes(expression.$kind)) {
+        expression = (expression as ValueConverterExpression | BindingBehaviorExpression).expression;
+        observingInnerItems = true;
```
[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/runtime-html/src/resources/template-controllers/repeat.ts]: You are right. Moved it to `binding`. Thanks!

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 3]
```diff
@@ -156,13 +182,29 @@ export class Repeat<C extends Collection = unknown[]> implements ICustomAttribut
...
+      }
+      this.innerItemsExpression = expression;
+    }
+    let innerItems = this.innerItems;
+    if (observingInnerItems) {
+      innerItems = this.innerItems = this.innerItemsExpression?.evaluate(flags, scope, this.forOfBinding.locator, null) as Items<C> ?? null;
+      observingInnerItems = this.observingInnerItems = !Object.is(this.items, innerItems);
```
[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/runtime-html/src/resources/template-controllers/repeat.ts]: Changes done. Thanks!

File: packages/__tests__/3-runtime-html/repeat.vc.bb.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bigopon) [reply packages/__tests__/3-runtime-html/repeat.vc.bb.spec.ts:142]: For methods, they can also be floating fns that take an extra param or two. Though readability is probably the matter of opinion so let's keep things the way they are.

**Review by bigopon — APPROVED** [review]

File: packages/runtime-html/src/resources/template-controllers/repeat.ts [hunk 4]
```diff
@@ -156,13 +190,22 @@ export class Repeat<C extends Collection = unknown[]> implements ICustomAttribut
...
+    }
+
     const oldObserver = this._observer;
     if ((flags & LF.fromUnbind)) {
       if (oldObserver !== void 0) {
         oldObserver.unsubscribe(this);
       }
     } else if (this.$controller.isActive) {
-      const newObserver = this._observer = getCollectionObserver(this.items);
+      const newObserver = this._observer = getCollectionObserver(observingInnerItems ? innerItems : this.items);
```
[the remark this answers is not in the document]
Comment (Sayan751) [reply packages/runtime-html/src/resources/template-controllers/repeat.ts:208]: I thought in the line as well. But I want the existing infra of indexmap to work when the repeater is not wrapped with VC or BB.

**Pull request merged**